### PR TITLE
Add documentation for caching in a module

### PR DESCRIPTION
**Pull request by babolivier**
As of https://github.com/matrix-org/synapse/pull/13667 we have a proper way for module to cache stuff, but we don't have great documentation with it (or how all of the pieces fit together). So here's an attempt at documenting the current state of things.

### CODE REVIEW

File: docs/modules/writing_a_module.md
Comment (DMRobertson) [inline docs/modules/writing_a_module.md:93]: `@cached` has sharp edges. I would like to see a warning here or at least some kind of pointer to the docstring of `@cached`... which doesn't exist.

Are there limitations on functions that can be cached? IIRC they must not have any variadic args (no `*args, **kwargs`). When should a module author use `tree=True` or `iterable=True`?

Comment (babolivier) [reply docs/modules/writing_a_module.md:93]: I'm not sure I'm best positioned to be documenting `cached` given I haven't used it with arguments in a while. Presumably in most cases just using it without arguments should be fine. Do we want to block this on `cached` getting a docstring?

Comment (DMRobertson) [reply docs/modules/writing_a_module.md:93]: > Do we want to block this on `cached` getting a docstring?

Not especially. Back in the queue to solicit other opinions

File: docs/modules/writing_a_module.md
Comment (erikjohnston) [inline docs/modules/writing_a_module.md:90]:
```suggestion
can be helpful for modules that need to repeatedly request the same data from the database
```

File: docs/modules/writing_a_module.md
Comment (erikjohnston) [reply docs/modules/writing_a_module.md:93]: I'm almost inclined to really strip down the public API we expose for `@cached`. I understand why we exposed it to modules, but I'm *very* cautious about making all the various options a public API.

**Review by DMRobertson — DISMISSED** [review]
Marking as required changes so Brendan will see Erik's comment

**Comment by babolivier** [comment]
Despite saying otherwise to the team last month, I won't be able to work on this PR anymore, most of all because (as I kinda forgot) being made redundant by Element means I've lost my push access to the repo and I can therefore not push to this PR anymore.

Technically I could fork and reopen the PR, but I think at this point it's mostly there and maybe just needs a link to the module API helper that #14663 introduced (but should otherwise be fine to merge). So it sounds like less effort to just have someone from the team take the conn on it.

**Comment by DMRobertson** [comment]
> Despite saying otherwise to the team last month, I won't be able to work on this PR anymore, most of all because (as I kinda forgot) being made redundant by Element means I've lost my push access to the repo and I can therefore not push to this PR anymore.
> 
> Technically I could fork and reopen the PR, but I think at this point it's mostly there and maybe just needs a link to the module API helper that #14663 introduced (but should otherwise be fine to merge). So it sounds like less effort to just have someone from the team take the conn on it.

I have done as much, adding to the review queue.

File: docs/modules/writing_a_module.md
Comment (clokep) [reply docs/modules/writing_a_module.md:95]: Looks like this thread was done, so I'm going to resolve it.

File: docs/modules/writing_a_module.md
Comment (clokep) [inline docs/modules/writing_a_module.md:135]: I find this example slightly confusing because `get_user` doesn't exist?

```suggestion
    @cached()
    async def get_department_for_user(self, user_id: str) -> str:
        """A function with a cache."""
        # Request a department from an external service.
        return await self.http_client.get_json(
            "https://int.example.com/users", {"user_id": user_id)
        )["department"]

    async def do_something_with_users(self) -> None:
        """Calls the cached function and then invalidates an entry in its cache."""
        
        user_id = "@alice:example.com"
        
        # Get the user. Since get_department_for_user is wrapped with a cache,
        # the return value for this user_id will be cached.
        department = await self.get_department_for_user(user_id)
        
        # Do something with `department`...
        
        # Let's say something has changed with our user, and the entry we have for
        # them in the cache is out of date, so we want to invalidate it.
        await self.api.invalidate_cache(self.get_department_for_user, (user_id,))
```

**Review by clokep — APPROVED** [review]
The suggested change isn't required, but I found it a bit confusing.

**Pull request merged**
